### PR TITLE
sys/xtimer: improve spin define handling/documentation

### LINKED ISSUE #11522: doc: XTIMER_BACKOFF inconsistency
### Description

Looking into XTIMER_BACKOFF documentation I've run upon some inconsistency over the meaning of the value.

In multiple places XTIMER_BACKOFF is refered to as having usec units, eg:

https://github.com/RIOT-OS/RIOT/blob/568a3514557fe301aaf8b2413b76c701dac98492/sys/include/xtimer.h#L474-L484

https://github.com/RIOT-OS/RIOT/blob/568a3514557fe301aaf8b2413b76c701dac98492/sys/include/xtimer.h#L136-L145

But it is also referred  as ticks:

https://github.com/RIOT-OS/RIOT/blob/568a3514557fe301aaf8b2413b76c701dac98492/sys/include/xtimer.h#L160-L169

But when looking at where it is used, e.g tracing ` xtimer_usleep()`. We can actually see that the value is used a ticks since it is converted from usec to ticks inmediatly.

```
static inline void xtimer_usleep(uint32_t microseconds)
{
    _xtimer_tsleep32(_xtimer_ticks_from_usec(microseconds));
}
```

And when tracing the value from `_xtimer_tsleep32() `to  `_xtimer_tsleep()` to `_xtimer_set64() ` to `_xtimer_set()` we end up seeing that XTIMER_OFFSET is compared against ticks.

https://github.com/RIOT-OS/RIOT/blob/568a3514557fe301aaf8b2413b76c701dac98492/sys/xtimer/xtimer_core.c#L136-L155

This leads to unclear definition of XTIMER_BACKOFF since documentation makes you think of usec and not ticks. This has an impact on some tests, eg https://github.com/RIOT-OS/RIOT/blob/568a3514557fe301aaf8b2413b76c701dac98492/tests/ps_schedstatistics/main.c#L48

Where a variable representing ticks is passed to a function requesting usec. This had already been brought up in #8969.

### Proposed steps

- [ ] change documentation to make sure XTIMER_BACKOFF has a clear unit ( ticks or usec), or add XTIMER_BACKOFF_US
- [ ] fix tests that use XTIMER_BACKOFF as an unproper reference ([tests/ps_schechstatistics](https://github.com/RIOT-OS/RIOT/tree/master/tests/ps_schedstatistics), [tests/thread_float](https://github.com/RIOT-OS/RIOT/tree/master/tests/thread_float), etc.)
- [ ] fix XTIMER_BACKOFF definition for concerned boards [frdmk-kw41z](https://github.com/RIOT-OS/RIOT/blob/master/boards/frdm-kw41z/include/board.h)

#### Related Issues

#8969

### PR DESCRIPTION
**Pull request by SemjonKerner**
### Contribution description

This PR adds missing type conversion/documentation of multiple `XTIMER` defines, tackling confusion between hardware ticks and microseconds.

This PR also deletes `XTIMER_MIN_SPIN`, which by my digging was never used from first definition on, yet touched a lot... rabbit hole:
https://github.com/RIOT-OS/RIOT/commit/a719e7d61ff90ce3c40a6508a5ed5af6b873cfe9#diff-2fb2297f4238db71ba10656337cfe96eR451
https://github.com/RIOT-OS/RIOT/commit/f86c118594dd15447f67e1fd0162e492e3a733d6#diff-2fb2297f4238db71ba10656337cfe96eR458
https://github.com/RIOT-OS/RIOT/commit/856c5e1539a356cb9a7f1620caa365be59e88d51#diff-2fb2297f4238db71ba10656337cfe96eR463
https://github.com/RIOT-OS/RIOT/commit/63e91151cd0b3dfde28bd7cc9b3ef23b140f3730#diff-4cf746d1ae899bdb66748fb93ac17eb0R81

If you know of more broken definitions in XTIMER, please let me so I can add them.

### Testing procedure

Do some tests/xtimer_* and make sure the values are still sane with this PR.

### Issues/PRs references

Fixes #11522

### CODE REVIEW

**Comment by kaspar030** [comment]
LGTM, don't have time for any testing, though.

**Comment by SemjonKerner** [comment]
> LGTM, don't have time for any testing, though.

I am fine with that, maybe @fjmolinas has more time for that?

**Comment by cladmi** [comment]
I am currently running `tests/posix_semaphore tests/posix_time tests/pthread_barrier tests/pthread_cleanup tests/pthread_condition_variable tests/pthread_cooperation tests/pthread_rwlock tests/pthread_tls tests/xtimer_drift tests/xtimer_hang tests/xtimer_longterm tests/xtimer_msg tests/xtimer_msg_receive_timeout tests/xtimer_now64_continuity tests/xtimer_periodic_wakeup tests/xtimer_remove tests/xtimer_reset tests/xtimer_usleep tests/xtimer_usleep_short` tests on some boards.

I see this already breaks some `arduino-mega2560` tests.

* `tests/xtimer_periodic_wakeup` that gets stuck around `2019-05-20 17:06:38,792 - INFO # Testing interval 107... (now=5915664)`.
* `tests/xtimer_usleep` but was already broken since `stdio_uart_rx`
* `tests/xtimer_usleep_short` `2019-05-20 17:09:06,204 - INFO # going to sleep 90 usecs...
xtimer stuck when trying to sleep 90 usecs`

I will give more update on broken tests when I have results.

**Comment by cladmi** [comment]
I think all boards that define `XTIMER_HZ` different from 1000000 will need to have their `XTIMER_BACKOFF` updated.

**Comment by cladmi** [comment]
`frdm-kw41z` fails on the same 3 tests from that selection.

**Comment by cladmi** [comment]
Somehow the internal functions in ticks should be using the `xtimer_ticks32_t` and `xtimer_ticks64_t` types instead of just ignoring them. Not sure if it fits in this pull request too or not.

File: sys/xtimer/xtimer_core.c [hunk 1]
```diff
@@ -198,7 +198,7 @@ int _xtimer_set_absolute(xtimer_t *timer, uint32_t target)
     DEBUG("timer_set_absolute(): now=%" PRIu32 " target=%" PRIu32 " offset=%" PRIu32 "\n",
           now, target, offset);
 
-    if (offset <= XTIMER_BACKOFF) {
+    if (offset <= xtimer_ticks_from_usec(XTIMER_BACKOFF).ticks32) {
```
Comment (fjmolinas) [inline sys/xtimer/xtimer_core.c]: Can you use a define for this? This will always be a constant value right?

Comment (cladmi) [reply sys/xtimer/xtimer_core.c]: I think it will be optimized already anyway. Its a static inline function.

**Comment by cladmi** [comment]
I have failed tests on `tests/posix_time` `tests/xtimer_periodic_wakeup` `tests/xtimer_usleep` and `xtimer_usleep_short` with the boards

* arduino-mega2560
* frdm-kw41z
* msba2 (also fails on `pthread_tls` but was already before)
* nucleo-f103rb
* stlb001a

And no issues for tests from https://github.com/RIOT-OS/RIOT/pull/11546#issuecomment-494033710 on

* frdm-k64f
* mulle
* nrf52dk
* pba-d-01-kw2x
* stm32f3discovery

The raw output for the tests run. It is rebased on https://github.com/RIOT-OS/RIOT/pull/10870 to allow testing

https://ci-ilab.imp.fu-berlin.de/job/experimental-pull-request-tests/19/consoleText

And build artifacts https://ci-ilab.imp.fu-berlin.de/job/experimental-pull-request-tests/19/artifact/build/output/results/*zip*/results.zip

File: sys/xtimer/xtimer_core.c (hunk 1 above)
Comment (SemjonKerner) [reply sys/xtimer/xtimer_core.c]: XTIMER_BACKOFF is independently defined for some boards, as well as the clocks so this shouldn't be constant for every possible instance, am i wrong?

Comment (fjmolinas) [reply sys/xtimer/xtimer_core.c]: > XTIMER_BACKOFF is independently defined for some boards

Yes, but still constant value. Anyway it was more a question about compiler optimization but as @cladmi pointed out it should be optimized out sinze it is inline. So I'll resolve de conversation.

**Comment by SemjonKerner** [comment]
Thanks for testing @cladmi. I reproduced the failures on nucleo-f103rb for `tests/xtimer_periodic_wakeup` and `xtimer_usleep_short`. I reconfigured the boards config for `XTIMER_BACKOFF` in `boards/common/nucleo/include/board_nucleo.h` for STM32F1 family from **5** to **19** (confirmed as smallest working with the broken tests).

So I guess, this would be the correct approach to solve the issues for all boards that redefine `XTIMER_HZ`.

Question is now, is it worth the effort to reevaluate best fitting backoff values or is this PR in vain and XTIMER_BACKOFF should be used as ticks instead, as it seems to be used as ticks in most cases?

**Comment by fjmolinas** [comment]
> Question is now, is it worth the effort to reevaluate best fitting backoff values or is this PR in vain and XTIMER_BACKOFF should be used as ticks instead, as it seems to be used as ticks in most cases?

I thin your comments in #11522 where sensible, `XTIMER_BACKOFF` as `us` is more user-friendly. Changing to us does seem like the more tedious way to got but the best in the end. Only board defining `XTIMER_HZ != `1000000` should be affected by your changes. I can provide the changes for frdm-kw41z (which I have), but I don't have any of the other boards.

For the other ones we could just manually re-calculate those values and it should work, as long as  xtimer_ticks_from_usec(XTIMER_BACKOFF).ticks32) = XTIMER_BACKOFF (before this PR) it should still work (or work just as before).

Anyways, yes its more work but I thinks its a sensible change, so +1 for this PR on my side. (I'll try to make time next week to help with testing on the board I have)

**Comment by SemjonKerner** [comment]
So, I crawled all boards that touch XTIMER_BACKOFF or XTIMER_HZ and just started a list with needed changes to XTIMER_BACKOFF. Since I don't have all these boards I need your help.

I'd be happy if you help to find optimal backoff values (make sure to use this PR for testing).
For some board change the backoff value and test with all tests that were used by @cladmi until an optimal value is found.
Then mark the board and write down the new XTIMER_BACKOFF value. You may leave a brief note. In case no change is needed cross the board with ~~. If you lack rights to edit my comment summon me.

XTIMER_BACKOFF && XTIMER_HZ changed:
- [x] common/arduino-atmega _`XTIMER_BACKOFF (78)` on arduino-mega2560_
- [ ] frdm-kw41z
- [ ] mega-xplained
~~mulle~~ _unused definitions ignored, see: https://github.com/RIOT-OS/RIOT/blob/b4fbfd7a7c25f13f44cecfc22627aec5d7c26baf/boards/mulle/include/board.h#L44_
- [ ] phynode-kw41z
- [ ] usb-kw41z
- [ ] waspmote-pro

only XTIMER_HZ changed:
- [ ] hifive1
- [ ] ikea-tradfri
- [ ] jiminy-mega256rfr2
- [ ] slstk3401a
- [ ] slstk3402a
- [ ] sltb001a
- [ ] slwstk6000b
- [ ] stk3600
- [ ] stk3700

only XTIMER_BACKOFF changed:
- [ ] blackpill
- [ ] bluepill
- [ ] cc2538dk
- [ ] cc2650-launchpad
- [ ] cc2650stk
- [ ] common/esp866
- [ ] common/msb-430
- [ ] common/nrf51
- [ ] common/nucleo _`XTIMER_BACKOFF (19)` stm32f1 on nucleo-f103rb_
- [ ] common/nucleo _stm32l1_
- [ ] common/nucleo _stm32f4_
- [ ] common/remote
- [ ] common/wsn430
- [ ] lobaro-lorabox
- [ ] maple-mini
- [ ] opencm904
- [ ] openmote-b
- [ ] openmote-cc2538
~~stm32f4discovery~~
- [ ] teensy31
- [ ] telosb
- [ ] z1

**Comment by fjmolinas** [comment]
Tested for kw41z (on frdm-kw41z but should be the same for all kw41z boards).  There are still 2 tests failing but this has to do with the fact that `xtimer_ticks_from_usec()` rounds downs so it sleeps less than the target time and we get negative offsets.

```
ERROR:frdm-kw41z:Tests failed: 2
Failures during test:
- [tests/xtimer_periodic_wakeup](tests/xtimer_periodic_wakeup/test.failed)
- [tests/xtimer_usleep](tests/xtimer_usleep/test.failed)

```

**Comment by fjmolinas** [comment]
> only XTIMER_BACKOFF changed:

These don't need to be changed as their timer speed is 1Mhz, 1tick=1usec. Although maybe they can be improved, I wouldn't target this for this pr.

> I have failed tests on tests/posix_time tests/xtimer_periodic_wakeup tests/xtimer_usleep and xtimer_usleep_short with the boards
> 
> nucleo-f103rb

@cladmi I find it weird that nucleo-f103rb is failing, the changes of this PR should not impact this board. Are you sure it wasn't failing in master?

**Comment by SemjonKerner** [comment]
@fjmolinas I compiled nucleo-f103rb myself and had the same issue. I found a working backoff, so these boards might be affected as well. However I had the same thought and hence divided the groups in the first place.

**Comment by fjmolinas** [comment]
@SemjonKerner and those tests where passing on master? (sorry I didn't understand if you where able to reproduce on your branch and/or master) If thats the case then you are right, we will have to re-check all cases.

**Comment by SemjonKerner** [comment]
> @SemjonKerner and those tests where passing on master? (sorry I didn't understand if you where able to reproduce on your branch and/or master) If thats the case then you are right, we will have to re-check all cases.

You are right. The problem occurs on master as well, for the nucleo-f103rb! Tests `xtimer_periodic_wakeup` and `xtimer_usleep_short` will not conclude. Even changing the backoff in master works as easily as it does in this PR.

Yet for all boards only changing `XTIMER_HZ` there should be a new define for `XTIMER_BACKOFF`, except for those were the default value is correct though.

**Comment by cladmi** [comment]
Would it make sense to do it the "easy way" with a new variable `XTIMER_BACKOFF_US`, and deprecate `XTIMER_BACKOFF` and when it is not defined we keep the old behavior?

This way, we can migrate the code without modifying all boards as it does not break the current behavior.

Then updating the boards can be done incrementally.

One issue I see in the low level code, is that not even `xtimer` is using `xtimer_spin(xtimer_ticks32_t)`, replacing to use it would have prevented these issues and the conversion to `_xtimer_spin(ticks.ticks32)` would be optimized by the compiler anyway.

**Comment by SemjonKerner** [comment]
> Would it make sense to do it the "easy way" with a new variable `XTIMER_BACKOFF_US`, and deprecate `XTIMER_BACKOFF` and when it is not defined we keep the old behavior?

I can imagine how this would work, yet I personally don't feel comfortable doing that.

You mean like
```
#ifndef XTIMER_BACKOFF_US

if (offset < xtimer_ticks_from_usec(XTIMER_PERIODIC_SPIN).ticks32) { ... }

#else

if (offset < XTIMER_PERIODIC_SPIN) { ... }

#endif
```
And then, for all boards that we suspect are affected by this change, we set the `XTIMER_BACKOFF_US` to the boards old `XTIMER_BACKOFF`?

This is an option, but I prefer doing it right away for sake of clean code. Also since the old behavior is only broken in some corner cases I would prefer to give this PR two more weeks and make it right.

**Comment by cladmi** [comment]
In this case I would also put a `XTIMER_PERIODIC_SPIN_US` and a `XTIMER_PERIODIC_RELATIVE_US` too.

```
#ifdef XTIMER_PERIODIC_SPIN_US

if (offset < xtimer_ticks_from_usec(XTIMER_PERIODIC_SPIN_US).ticks32) { ... }

#else
/* Compatibility, XTIMER_PERIODIC_SPIN will be deprecated */

if (offset < XTIMER_PERIODIC_SPIN) { ... }

#endif
```

> This is an option, but I prefer doing it right away for sake of clean code. Also since the old behavior is only broken in some corner cases I would prefer to give this PR two more weeks and make it right.

This changes the behavior instantly and breaks immediately boards outside and pull requests that are still using the previous values.

For clean code right away, the values in ticks should be updated to use the `ticks` type instead on just `uint32_t` too.

**Comment by SemjonKerner** [comment]
Okay, since this PR is not moving by itself and I dont have any more boards from the list I will close it and reopen a PR with the suggested approach by @cladmi soonish.

**Pull request closed without merging**